### DISCUSSION TITLE
I need to find the dominance tree or a dominance of a program function

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#3707] [Q&A] [answered]
There is a `'''public static <V,​E extends GEdge<V>> GDirectedGraph<V,​GEdge<V>> ` API on Graph Algorithm. Is it possible to get the dominance tree in headless? it says V,E of the graph, what are these? Can a function dominance be found out by the following API, or basicblock's dominance tree? 
Oh man! I'm probably not gonna get answer for this!

**Comment by dragonmacher** [comment]
You should be able to use the API you mentioned after first creating a graph of all functions, with edges connecting functions via calls (it should work in headless).  

The 'V,E' represent 'vertex' and 'edge' type of your choosing.   In your case, the V is a ghidra `Function` object.   The edge can be one of our `DefaultGEdge`s, with a V of `Function`.

Most of our code uses dominance based on code blocks.  For an example, see the `DominanceModularizationCmd`:
```
	@Override
	protected void applyModel() throws CancelledException {
		GDirectedGraph<CodeBlockVertex, CodeBlockEdge> callGraph = createCallGraph();
		CodeBlockVertex root = createRoot(callGraph, "Dominance");

		GDirectedGraph<CodeBlockVertex, GEdge<CodeBlockVertex>> dominanceGraph =
			GraphAlgorithms.findDominanceTree(callGraph, monitor);

		rebuildProgramTree(dominanceGraph, root, destinationModule);
	}
```

**Reply by anzosasuke** [reply]
thanks for helping out. I think making dominance tree for codeblock is more versatile. Where you used code blocks, is that same as code blocks from BasicBlockModel? If vertex is basic block, what is the edge? is it edge from one basic block to other block? I was waiting for the Ghidra API website to work again, seems like it has been down for a week.

**Reply by dragonmacher** [reply]
> Where you used code blocks, is that same as code blocks from BasicBlockModel?

Yes, or any block model you wish to use.

> If vertex is basic block, what is the edge? is it edge from one basic block to other block?

Yes.  The edges simply represent the flow between code blocks.

**Reply by anzosasuke** [reply]
it seems that ghidra.re webistes are still under maintenance. I tried this. I found there is Dominator class in DirectedGraph.
I have created graph(v,e) called digraph.

```
dominator = Dominator(dirgraph)
graph = dominator.getDominanceGraph()
vertices = graph.vertexIterator()
while vertices.hasNext():
    vertex = vertices.next()
    dom = graph.getDominator(vertex) 
    print("{} has dominator {}".format(vertex, dom))```

BUt I am not getting the dominator. the output looks like this. How do I get the dominator of particular vertex?

```FUN_00101020 has dominator FUN_00101020
dont_do_this has dominator dont_do_this
FUN_00101020 has dominator FUN_00101020
innocuous has dominator innocuous
FUN_00101020 has dominator FUN_00101020
```

**Reply by dragonmacher** [reply]
The `Dominator` class is no longer used in Ghidra and should be deprecated.  Unfortunately, I have no sense of how well it works.

My suggestion is to get the Ghidra source code and use the code I mentioned in my previous example as a guide to get started.  That code is tested and uses Graph APIs that should work and have other example usage throughout the code.

**Reply by anzosasuke** [reply]
About that GDirectedGraph, I am having trouble creating that graph as I don't know how to create the GDirectedGraph object?
There is no constructor object, I am trying to create it with no luck. I use python, Is there way I can use to create that GDirectedGraph object.
Like for example:
```
dirgraph = DirectedGraph() 
digraph.add(Vertex(func)) 

```

I can't seem to accomplish same for GDirectedGraph?

**Reply by dragonmacher** [reply]
My Python skills are virtually non-existent.   This is the Java code snippet from the code I recommended that creates the graph:

```
GDirectedGraph<CodeBlockVertex, CodeBlockEdge> graph = GraphFactory.createDirectedGraph();
```

The `GraphFactory` creates a directed graph for you via the static method call.

**Reply by anzosasuke** [reply]
I tried creating the Gdirected Graph using above code. It works up to a point but the graph created was JungDirectedGraph.
My code 
```
digraph = GraphFactory.createDirectedGraph()
listing = currentProgram.getListing()
fm = currentProgram.getFunctionManager()

funcs = fm.getFunctions(True) # True mean iterate forward
for func in funcs: 
    # Add function vertices
    print("Function: {} @ 0x{}".format(func.getName(), func.getEntryPoint())) # FunctionDB
    digraph.addVertex(func)
    entryPoint = func.getEntryPoint()
    instructions = listing.getInstructions(entryPoint, True)
    for instruction in instructions:
        addr = instruction.getAddress()
        oper = instruction.getMnemonicString()
        if oper == "CALL":
            print("    0x{} : {}".format(addr, instruction))
            flows = instruction.getFlows()
            if len(flows) == 1:
                target_addr = "0x{}".format(flows[0])
                edge = DefaultGEdge(func, fm.getFunctionAt(getAddress(target_addr)))
                digraph.addEdge(edge)

gdgraph = digraph.copy()
callgraph = gdgraph.createCallGraph()
root = createRoot(callgraph, "Dominance")

dominanceGraph = GraphAlgorithms.findDominanceTree(callgraph, monitor)```

This is giving me this error.

 callgraph = gdgraph.createCallGraph()
AttributeError: 'ghidra.graph.jung.JungDirectedGraph' object has no attribute 'createCallGraph' 
 Is there any way to fix this  issue?

**Reply by dragonmacher** [reply]
There is no method `createCallGraph()` on the directed graph.  You should be able to pass `gdgraph` into `GraphAlgorithms.findDominanceTree()` directly.

**Reply by anzosasuke** [reply]
I did that.
 `graph = GraphAlgorithms.findDominanceTree(digraph, monitor)`
But I got this error.
```line 48, in <module>
    graph = GraphAlgorithms.findDominanceTree(digraph, monitor)
	at edu.uci.ics.jung.graph.util.Pair.<init>(Pair.java:42)
	at edu.uci.ics.jung.graph.AbstractGraph.addEdge(AbstractGraph.java:60)
	at edu.uci.ics.jung.graph.AbstractGraph.addEdge(AbstractGraph.java:55)
	at ghidra.graph.jung.JungDirectedGraph.addEdge(JungDirectedGraph.java:27)
	at ghidra.graph.algo.ChkDominanceAlgorithm.getDominanceTree(ChkDominanceAlgorithm.java:236)
	at ghidra.graph.GraphAlgorithms.findDominanceTree(GraphAlgorithms.java:306)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Pair cannot contain null values```

**Reply by dragonmacher** [reply]
Based on that trace, it seems like there is a flaw in the algorithm.  It is attempting to create an edge where at least one of the vertices for the edge is null.   It looks like `V dominator = getImmediateDominator(v);` may return null, which this code is not checking for.

**Reply by anzosasuke** [reply]
So, Should I make some changes? how do I fix this? Is something wrong with the code?

**Reply by dragonmacher** [reply]
> So, Should I make some changes? how do I fix this? Is something wrong with the code?

My guess is that the graph fed to the algorithm is created in a way that allowed this to happen.  I can't say for sure without being able to debug the issue.   My guess is that you added a `null` vertex to an edge you created in this code:
```
 edge = DefaultGEdge(func, fm.getFunctionAt(getAddress(target_addr)))
```

This `fm.getFunctionAt(getAddress(target_addr))` can return null.

**Reply by anzosasuke** [reply]
So i have made basicblockmodel's Gdirected graph and fed it into the findDominanceTree() from GraphAlgorithms. But I don't know why the output is not what it should be. This is my whole code.

```from ghidra.program.model.block.graph import CodeBlockVertex
from ghidra.program.model.block.graph import CodeBlockEdge
from ghidra.graph import GraphAlgorithms
from ghidra.app.cmd.module import DominanceModularizationCmd
from ghidra.program.model.block import BasicBlockModel
from ghidra.graph import GraphFactory


    
digraph = GraphFactory.createDirectedGraph()

bbm = BasicBlockModel(currentProgram)
blocks = bbm.getCodeBlocks(monitor)
block = blocks.next()

while block:
	vert = CodeBlockVertex(block)
	digraph.addVertex(vert)

	block = blocks.next()
	vert_next = CodeBlockVertex(block)
	edge = CodeBlockEdge(vert, vert_next)
	digraph.addEdge(edge)
print(digraph)
dominancegraph = GraphAlgorithms.findDominanceTree(digraph, monitor)
print(graph)
from_vertex = "innocuous"
dom = GraphAlgorithms.findDominance(digraph, from_vertex ,monitor)
print(dom)
```

And the output I get is same with digraph and dominancegraph. To get the dominance right, what am i missing? please help me.

**Comment by dragonmacher** [comment]
You appear to no be creating the edges correctly.  You are not using the destinations of the code blocks to create edges.  As I referenced before, you can see an example of this in `DominanceModularizationCmd` in the `addEdgesForDestinations()` method:
```
private void addEdgesForDestinations(GDirectedGraph<CodeBlockVertex, CodeBlockEdge> graph,
			CodeBlockVertex fromVertex, CodeBlock sourceBlock,
			Map<CodeBlock, CodeBlockVertex> instanceMap) throws CancelledException {

		CodeBlockReferenceIterator iterator = sourceBlock.getDestinations(monitor);
		while (iterator.hasNext()) {
			monitor.checkCanceled();

			CodeBlockReference destination = iterator.next();
			CodeBlock targetBlock = getDestinationBlock(destination);
			if (targetBlock == null) {
				continue; // no block found
			}

			CodeBlockVertex targetVertex = instanceMap.get(targetBlock);
			if (targetVertex == null) {
				targetVertex = new CodeBlockVertex(targetBlock);
				instanceMap.put(targetBlock, targetVertex);
			}

			graph.addVertex(targetVertex);
			graph.addEdge(new CodeBlockEdge(fromVertex, targetVertex));
		}
	}
```

I suggest use use that class as a guide to setup your graph.

**Reply by anzosasuke** [reply]
Thanks. I made my graph according to above mentioned class. This is my whole script.
```
from ghidra.program.model.block.graph import CodeBlockVertex
from ghidra.program.model.block.graph import CodeBlockEdge
from ghidra.graph import GraphAlgorithms
from ghidra.program.model.block import BasicBlockModel
from ghidra.graph import GraphFactory
from ghidra.graph import GDirectedGraph
from ghidra.program.model.address import *
from ghidra.program.model.block import *
from ghidra.program.model.listing import *
from ghidra.program.util import GroupPath


def getDestinationBlock(destination):
	targetAddress = destination.getDestinationAddress()
	targetBlock = bbm.getFirstCodeBlockContaining(targetAddress, monitor)
	if targetBlock == None:
		return None
	blockAddress = targetBlock.getFirstStartAddress()
	return targetBlock

digraph = GraphFactory.createDirectedGraph()

bbm = BasicBlockModel(currentProgram)
blocks = bbm.getCodeBlocks(monitor)

while blocks.hasNext():

	block = blocks.next()
	fromVertex = CodeBlockVertex(block)
	digraph.addVertex(fromVertex)
	iterator = block.getDestinations(monitor)
	while (iterator.hasNext()):
		destination = iterator.next()
		print(destination)
		targetBlock = getDestinationBlock(destination)
		print(targetBlock)
		if targetBlock == None:
			continue
		targetVertex = CodeBlockVertex(targetBlock)

		digraph.addVertex(targetVertex)
		edge = CodeBlockEdge(fromVertex, targetVertex)
		digraph.addEdge(edge)

print(digraph)
graph = GraphAlgorithms.findDominanceTree(digraph, monitor)
print(graph)
from_vertex = "innocuous"
dom = GraphAlgorithms.findDominance(digraph, from_vertex ,monitor)
print(dom)
```

But I get the error which i got before:

```graph = GraphAlgorithms.findDominanceTree(digraph, monitor)
	at edu.uci.ics.jung.graph.util.Pair.<init>(Pair.java:42)
	at edu.uci.ics.jung.graph.AbstractGraph.addEdge(AbstractGraph.java:60)
	at edu.uci.ics.jung.graph.AbstractGraph.addEdge(AbstractGraph.java:55)
	at ghidra.graph.jung.JungDirectedGraph.addEdge(JungDirectedGraph.java:27)
	at ghidra.graph.algo.ChkDominanceAlgorithm.getDominanceTree(ChkDominanceAlgorithm.java:236)
	at ghidra.graph.GraphAlgorithms.findDominanceTree(GraphAlgorithms.java:306)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Pair cannot contain null values
 ```
Is there a way to get over this problem>??

**Comment by dragonmacher** [comment, marked as answer]
I do not see any obvious flaw in your script.   My guess is that the structure of your graph is such that you do not have a single source and/or a single sink in the graph.  The algorithm will create dummy source/sink vertices as needed in this case.   The algorithm appears to not handle the case where a dummy node is a dominator.   I am surpised by this, as I think we would have seen this already in our code.

Unfortunately, for you to work around this issue, you will have to change our source code.   You can do this if you setup a developer environment using our github repo.   Alternatively, you could copy our class(es) as needed, writing them in Python, which would allow you to fix the bug.  (Admittedly, this is not a nice solution.)

The fix we need is in `ChkDominanceAlgorithm` in this method:
```
	public GDirectedGraph<V, GEdge<V>> getDominanceTree() {
		GDirectedGraph<V, GEdge<V>> dg = GraphFactory.createDirectedGraph();

		// note: we use the source graph here and not the one we mutated for calculating dominance
		Collection<V> vertices = sourceGraph.getVertices();
		Set<V> sources = navigator.getSources(sourceGraph);
		for (V v : vertices) {
			if (sources.contains(v)) {
				continue;
			}

			V dominator = getImmediateDominator(v);
			if (!Objects.equals(dominator, v)) {
				dg.addEdge(new DefaultGEdge<>(dominator, v));
			}
		}
		return dg;
	}
```

Specifically, the code adding an edge needs to be changed to this:

```
	if (!Objects.equals(dominator, v) && dominator != null) {
		dg.addEdge(new DefaultGEdge<>(dominator, v));
	}
```

I intend to fix this, but I don't know how long it will be before you can access the updated code.

**Reply by dragonmacher** [reply]
In looking at how we use that code, it appears as though we ensure that the graphs we pass to the algorithm all have at least one source in the graph (a vertex with no incoming edges).   If you could manage to insert a fake source vertex into your graph, this would prevent the exception you are seeing.  Unfortunately, there is no easy way to do this given the code sample you have provided.

Edit: although not documented in our algorithm or utility, the algorithm requires at least once source vertex.  You may have multiple sources, but your graph must have at least one.  Otherwise, there is not way to find dominance.  You should be able to use `GraphAlgorithms.getSources()` to see if the graph has at least one source.  If it does not, then you cannot calculate dominance with that graph, at least not without modifying the graph first.

**Reply by anzosasuke** [reply]
Thanks so much. About the sources, I checked with my graph and I seem to have it(more than one). 
I think I will try changing the source code and then build it to check whether it works that way. If it doesn't, ahhhhh.... lets hope it does..

**Reply by anzosasuke** [reply]
Thanks so much. I was able to get over the error after changing the source code as you suggested. The graph I made had some vertices without edges which I am not sure why. I think that's why the error was triggered. I am still trying to figure out correct way to interpret the output though.

**Reply by dragonmacher** [reply]
You're welcome. I'm sure you will be able to figure out the issue with some debugging of how you create your vertices and edges from the code blocks.

**Reply by anzosasuke** [reply]
So I got the output for dominance tree. But i am perplexed on why the dominance tree graph and the Gdirected graph looks the same and I used the API `GraphAlgorithms.findDominance(digraph, fromVertex, monitor)` and the output is same as fromVertex everytime. I do not think the output is correct.

``` 

fromVertex = "get_count"
 dom = GraphAlgorithms.findDominance(digraph, fromVertex, monitor)
```

the output will be "get_count". Which should clearly be "main".

So i tried getting dominance of the basic block in a function. output snippets looks like this all the way similar.
```Function: FUN_00403444
GdirectedGraph
Vertices:FUN_00403444,FUN_00403450
Edges:[FUN_00403444, FUN_00403450][FUN_00403444,FUN_00403450] 
Dominance Tree
Vertices:FUN_00403444,FUN_00403450
Edges:[FUN_00403444, FUN_00403450][FUN_00403444,FUN_00403450] ```

**Reply by dragonmacher** [reply]
I believe the issue is that your example is too simple.   I suggest you look at the `ghidra.graph.GraphAlgorithmsTest`.  It has many examples of finding dominance with different graph setups.  As a bonus, it has some ascii depictions of the graphs.
